### PR TITLE
refactor and generalize transitions

### PR DESCRIPTION
**Pull request by boygirl**
cc/ @divmain 

This PR generalizes Transitions so that it can be used with lone components or with components that need to synchronize their children.

### CODE REVIEW

**Comment by boygirl** [comment]
cc/ @divmain @kenwheeler 

This PR:
- Generalizes the transitions so that they can work with single data components (i.e. bar), and nested and deeply nested data components (i.e. stacked bars, charts containing stacked bars...) 
- Adds `VictoryTransition` wrapper component which coordinates transition state, and animates its child.
- Adds another custom interpolator to replace `d3-interpolate.array` The change made here has also been proposed PR'd https://github.com/d3/d3-interpolate/pull/19  
- Needs moar tests, but I want to be certain I haven't overlooked something dumb first

File: src/victory-transition/victory-transition.jsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    this.setState(this.getTransitionState(this.props, nextProps));
+  }
+
+  getTransitionState(props, nextProps) {
+    const { animate } = props;
+    if (!animate) {
+      return {};
+    } else if (animate.parentState) {
```
Comment (divmain) [inline src/victory-transition/victory-transition.jsx:33]: Is `animate.parentState` necessary?  This seems harder to reason about than managing state from a single place.

File: src/victory-util/transitions.js [hunk 2]
```diff
@@ -70,28 +77,32 @@ export function getInitialTransitionState(oldChildren, nextChildren) {
...
 
     nodesWillExit = nodesWillExit || !!exiting;
     nodesWillEnter = nodesWillEnter || !!entering;
 
-    return { entering, exiting };
-  });
+    return { entering: entering || false, exiting: exiting || false };
+  };
+
+  const getTransitionsFromChildren = (old, next) => {
```
Comment (divmain) [inline src/victory-util/transitions.js:94]: These parameter names should be more descriptive.

File: src/victory-util/transitions.js [hunk 3]
```diff
@@ -175,51 +185,60 @@ function getChildPropsOnEnter(animate, data, enteringNodes) {
...
+  const parentAnimate = props.animate || {};
+  const parentDurations = {
     exit: parentAnimate.onExit && parentAnimate.onExit.duration || null,
     enter: parentAnimate.onEnter && parentAnimate.onEnter.duration || null,
     move: parentAnimate.duration || null
-  });
+  };
+
+  const children = React.Children.toArray(props.children);
+  if (children) {
```
Comment (divmain) [inline src/victory-util/transitions.js]: The docs don't say that `React.Children.toArray` will return a falsey value if `children` are falsey.  Is that the case?

**Comment by divmain** [comment]
I like that a generic wrapper could be pulled out to accommodate stand-alone components.  My only high-level feedback is that (for me) this seems a bit less explicit and a bit more "magic" - which may be what you intended, but is not my personal preference.  Mostly because I think it adds to maintenance costs.  That said, this looks like a very solid refactor!

**Comment by kenwheeler** [comment]
LGTM

**Pull request merged**